core/trino-main/src/main/java/io/trino/operator/project/MergePages.java

```diff
@@ -205,17 +208,27 @@ private void appendPage(Page page)
         {
             pageBuilder.declarePositions(page.getPositionCount());
             for (int channel = 0; channel < types.size(); channel++) {
-                appendBlock(
+                appendRawBlock(
                         types.get(channel),
-                        page.getBlock(channel).getLoadedBlock(),
+                        page.getBlock(channel),
                         pageBuilder.getBlockBuilder(channel));
             }
         }
 
-        private void appendBlock(Type type, Block block, BlockBuilder blockBuilder)
+        private void appendRawBlock(Type type, Block rawBlock, BlockBuilder blockBuilder)
         {
-            for (int position = 0; position < block.getPositionCount(); position++) {
-                type.appendTo(block, position, blockBuilder);
+            Block loadedBlock = rawBlock.getLoadedBlock();
+            int length = loadedBlock.getPositionCount();
+            switch (loadedBlock) {
+                case RunLengthEncodedBlock rleBlock -> blockBuilder.appendRepeated(rleBlock.getValue(), 0, length);
+                case DictionaryBlock dictionaryBlock -> {
+                    // TODO figure out why blockBuilder.appendPositions(dictionaryBlock.getDictionary(), dictionaryBlock.getRawIds(), 0, length); fails tests
+                    for (int position = 0; position < dictionaryBlock.getPositionCount(); position++) {
+                        type.appendTo(dictionaryBlock, position, blockBuilder);
+                    }
+                }
+                case ValueBlock valueBlock -> blockBuilder.appendRange(valueBlock, 0, length);
+                default -> throw new IllegalArgumentException("Unsupported block type " + loadedBlock.getClass().getSimpleName());
```

Comment:
Instead of adding default here, add an explicit case for LazyBlock, and then the case will be covering of all possible subclasses, so you won't need the default clause, and if we add a new type in the future, this code won't compile